kafkaesk/retry.py

```diff
@@ -0,0 +1,354 @@
+        ).time():
+            handler_key, handler = self._get_handler(exception)
+
+            if retry_history is None:
+                retry_history = RetryHistory()
+
+            # Add information about this failure to the history
+            retry_history.failures.append(
+                FailureInfo(
+                    exception=exception.__class__.__name__,
+                    handler_key=handler_key,
+                    timestamp=datetime.now(),
+                )
+            )
+
+            try:
+                await handler(self, handler_key, retry_history, record, exception)
+            finally:
+                RETRY_POLICY.labels(
+                    stream_id=record.topic,
+                    partition=record.partition,
+                    group_id=self.subscription.group,
+                    handler=handler.__class__.__name__,
+                    exception=exception.__class__.__name__,
+                ).inc()
+
+    def _get_handler(self, exception: Exception) -> Tuple[str, "RetryHandler"]:
+        exception_type = exception.__class__
+
+        handler_key, handler = self._handler_cache.get(exception_type, (None, None))
+
+        if handler is None:
+            handler = self._handlers.get(exception_type)
+            if handler is not None:
+                handler_key = exception_type.__name__
+                self._handler_cache[exception_type] = (handler_key, handler)
+
+        if handler is None:
+            for handler_exception_type in self._handlers.keys():
+                if isinstance(exception, handler_exception_type):
+                    handler = self._handlers[handler_exception_type]
+                    handler_key = handler_exception_type.__name__
+                    self._handler_cache[exception_type] = (handler_key, handler)
+                    break
+
+        # Return the default handler
+        if handler is None:
+            handler = self._default_handler
+
+        if handler_key is None:
+            handler_key = "Exception"
+
+        return (handler_key, handler)
+
+
+class RetryHandler(ABC):
+    """Base class implementing common logic for RetryHandlers
+
+    All RetryHandler's should implement the following metrics:
+
+    * RETRY_HANDLER_FORWARD - Note: This is implemented by RetryHandler._forward_message
+    * RETRY_HANDLER_DROP - Note: this is implemented by RetryHandler._drop_message
+    * RETRY_CONSUMER_TOPIC_OFFSET
+    * RETRY_CONSUMER_MESSAGE_LEAD_TIME - Note: If a RetryHandler's consumer expects a delay,
+        this delay should be subtracted from the lead time
+    * RETRY_CONSUMER_CONSUMED_MESSAGE_TIME
+    * RETRY_CONSUMER_CONSUMED_MESSAGES
+
+    See `kafkaesk.metrics` for more information on each of these metrics
+    """
+
+    logger = logging.getLogger("kafkaesk.retry.retry_handler")
+
+    def __init__(self) -> None:
+        self._ready = False
+
+    async def initialize(self, policy: RetryPolicy) -> None:
```

Comment:
Do we think we need the `initialize/finalize` interfaces here?

We could reduce a bunch of code if we just removed them. Maybe having access to the application and subscriber objects would provide everythign needed?

Reply:
So the `initialize` and `finalize` methods of `RetryHandler` aren't actually used by any of the current `RetryHandlers`. Still, I see them being used once we need to set up and manage consumers for a handler that is actually consuming some of these forwarded messages.

A more complicated `RetryHandler`, such as a `Backoff` implementation, will have some user-supplied configuration (retry count, retry delays) supplied through the constructor and will then need to perform some additional setup (creating consumers), which would be handled in the `initialize` method.

I looked at implementing 'lazy initialization' where we wouldn't run this code until the first time the handler is called, but if `initialize` handles setting up consumers, we could run into a scenario where we have retry messages available but are not processed because the lazy initialization was not triggered.

I also took a look at running this code from the constructor, but that ends up meaning we need to execute async code from the constructor, and we will need access to an event loop.  I felt that keeping all of the async code in an async `initialize` function simplifies the code and makes what is happening a little bit clearer to callers of the code.